src/storage-controller/src/persist_handles.rs

```diff
@@ -240,6 +240,65 @@ enum PersistTableWriteCmd<T: Timestamp + Lattice + Codec64> {
     Shutdown,
 }
 
+async fn append_work<T2: Timestamp + Lattice + Codec64>(
+    frontier_responses: &tokio::sync::mpsc::UnboundedSender<StorageResponse<T2>>,
+    write_handles: &mut BTreeMap<GlobalId, WriteHandle<SourceData, (), T2, Diff>>,
+    mut commands: BTreeMap<GlobalId, (tracing::Span, Vec<Update<T2>>, Antichain<T2>)>,
+) -> Result<(), Vec<GlobalId>> {
+    let futs = FuturesUnordered::new();
+
+    // We cannot iterate through the updates and then set off a persist call
+    // on the write handle because we cannot mutably borrow the write handle
+    // multiple times.
+    //
+    // Instead, we first group the update by ID above and then iterate
+    // through all available write handles and see if there are any updates
+    // for it. If yes, we send them all in one go.
+    for (id, write) in write_handles.iter_mut() {
+        if let Some((span, updates, new_upper)) = commands.remove(id) {
+            let persist_upper = write.upper().clone();
+            let updates = updates
+                .into_iter()
+                .map(|u| ((SourceData(Ok(u.row)), ()), u.timestamp, u.diff));
+
+            futs.push(async move {
+                let persist_upper = persist_upper.clone();
+                write
+                    .compare_and_append(updates.clone(), persist_upper.clone(), new_upper.clone())
+                    .instrument(span.clone())
+                    .await
+                    .expect("cannot append updates")
+                    .or(Err(*id))?;
+
+                Ok::<_, GlobalId>((*id, new_upper))
+            })
+        }
+    }
+
+    if !commands.is_empty() {
+        error!(
+            "Received appends for unregistered collections: {:?}",
+            commands.keys()
+        );
```

Comment:
It turns out this is reported during the cluster tests, which is a bit surprising:

```
cluster-materialized-1     | environmentd: 2023-10-23T21:51:34.344014Z  INFO advance_ready{otel.name="execute"}: mz_adapter::catalog::state: create table materialize.public.bigger (u13)
cluster-materialized-1     | environmentd: 2023-10-23T21:51:34.368627Z ERROR mz_storage_controller::persist_handles: Received appends for unregistered collections: [User(13)]
cluster-materialized-1     | environmentd: 2023-10-23T21:51:34.412727Z  INFO advance_ready{otel.name="execute"}: mz_storage_controller: advancing u13 to initial since of 1698097894379
```

I'll look a bit harder and figure out if this is scary or not.

Reply:
If it helps, I want to say that I found a race once in table creation, where the new table would show up in the list of things in group commit very slightly before it got registered with the table worker

Reply:
Yeah, that does seem relevant... if we change the implementation to fail these writes, all the failures are during group commit.